### DISCUSSION TITLE
Limiting (Remove) pyi_rth_win32comgenpy hook

### DISCUSSION THREAD

**Discussion by arossert** [pyinstaller/pyinstaller#8309] [Features & Development]
The purpose of this hook is to fix an issue with old versions of `pywin32` with creating the `gen_py` cache directory in the `site-packages`, nowadays this is not the case, this hook is always creating the `gen_py` file even if my app is not using `win32com`.

So my question is if we can remove this hook or limit the runtime to older versions of `pywin32`? 
Maybe we can add something like
```
if check_requirement('pywin32 >= ?'):
    _pyi_rthook()
    del _pyi_rthook
```
(I don't really know if these helper functions are available in runtime hooks, if not we can change it to something else)

I don't mind creating the PR if this solution is acceptable, I have some experience with adding hooks but not runtime hooks 😃

**Comment by rokm** [comment]
If your app isn't using `win32com`, wouldn't it be easier to just exclude the module (via `--exclude` or `excludes` argument to Analysis) in the first place?

As for the hook... it seems that the change in `pywin32` to use temporary directory was done a long time ago (https://github.com/mhammond/pywin32/commit/8fee3205c125a114cf2b171b4eb2e2137716fb2e), so perhaps we really should remove the hook altogether.

On the other hand, the logic for selection of `__gen_path__` still seems to contain paths that might end up favoring the package's directory. So just in case this is still a possibility, we might instead limit the hook by checking whether original `win32.__gen_path__` points inside `sys._MEIPASS` and if it does, relocate it. I think this would more surefire than checking for version (which I'm not sure we can obtain without collecting metadata, and might not be a good criterion if path selection is condition-based rather than version-based).

**Reply by arossert** [reply]
This sounds like a good solution, let me know if this is something I can do (not have much experience with the runtime hooks) 😃

**Reply by arossert** [reply]
@rokm Do you think that adding this to the hook can do the trick?
```
if Path(win32com.__gen_path__) in Path(sys._MEIPASS).parent:
    return
```

**Reply by rokm** [reply]
Off the top of my head, I think it should be

```python
# Check if `win32com.__gen_path__` is rooted in `sys._MEIPASS`
if pathlib.Path(sys._MEIPASS) in pathlib.Path(win32com.__gen_path__).parents:
   need_to_relocate = True
```

If you want to test it, you can freeze 
```
import win32com
print(win32com.__gen_path__)
```

in onedir, and run it with modified hook. It should use the default tmp directory. If you create an empty `_internal\win32com\gen_py` dir in frozen application's directory, it should relocate the tmp directory to our ephemeral one.

---

FWIW, I think I would also be fine with just removing the run-time hook - I suppose if user goes out of their way to collect `win32com\gen_py` directory into their frozen application, we might as well use that. 

If hook removal causes issues, we can reinstate it and limit it as discussed above. Otherwise, I suspect we're going to end up keeping that hook alive forever, "just in case".

@bwoodsend Which route would you prefer to take here?

**Reply by bwoodsend** [reply]
To be clear, it's [these two](https://github.com/mhammond/pywin32/blob/3e9da59af065a35485107a688e588690ffd72cdf/com/win32com/__init__.py#L102-L107) we never want to be used?

There isn't a `gen_py` directory in the (unfrozen) package so `__gen_py__` would as far as I can tell always eventually be set to `C:\\users\\brenainn\\Temp\\gen_py\\3.11`. In which case, most of that module and our hook working around it is dead code?

**Reply by rokm** [reply]
> To be clear, it's [these two](https://github.com/mhammond/pywin32/blob/3e9da59af065a35485107a688e588690ffd72cdf/com/win32com/__init__.py#L102-L107) we never want to be used?

Yeah. I think these two actually amount to the same thing, because if that `win32com\gen_py` directory existed, `win32com.genpy` would be a namespace package under contemporary python versions. So even those two codepaths themselves seem to be redundant (likely a left-over from times before PEP420 namespace packages).

> There isn't a `gen_py` directory in the (unfrozen) package so `__gen_py__` would as far as I can tell always eventually be set to `C:\\users\\brenainn\\Temp\\gen_py\\3.11`. In which case, most of that module and our hook working around it is dead code?

Indeed. If `__gen_py__`  already points to temporary directory, it makes little sense for us to force our ephemeral temporary directory for it (unless we specifically care that frozen applications use isolated, ephemeral temporary directories for that cache). 

We can choose between keeping the run-time hook but rendering it mostly no-op (by checking if original  `__gen_py__` is rooted in `sys._MEIPASS`), which gives us chance to revert to old behavior in case `win32com\gen_py` happens to exist for whatever reason. 

Or, simply remove the whole thing, and see if anyone complains; and only if they do, investigate that scenario and possibly re-instate the hook.

**Reply by bwoodsend** [reply]
Removing the whole thing sounds best to me.

**Reply by arossert** [reply]
@bwoodsend Should I create a PR?

**Reply by bwoodsend** [reply]
Sure, go for it!

**Reply by arossert** [reply]
Done
https://github.com/pyinstaller/pyinstaller/pull/8313
